### ISSUE TITLE
join a remote node by ddns

### ISSUE THREAD

**Issue by cailiangliang765** [kubernetes/kubeadm#1611]
I have created a kubernetes cluster where I have k8smaster1 and k8snode1. I initialised master node using below command

sudo kubeadm init --kubernetes-version=1.14.3 \
--apiserver-advertise-address=192.168.88.131 \
--image-repository registry.aliyuncs.com/google_containers \
--service-cidr=10.1.0.0/16 \
--pod-network-cidr=10.244.0.0/16

192.168.88.131 is the ip address of master node. Then I used the generated token to start my worker nodes. 
[root@k8snode1 ~]# kubeadm join 192.168.88.131:6443 --token idiny6.30idongoafemg1bp \
> --discovery-token-ca-cert-hash sha256:a9ac241ea191d51a99c311ba57265eeff6733298530243c1a0b1e143062b9e08
[preflight] Running pre-flight checks
[preflight] Reading configuration from the cluster...
[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[kubelet-start] Downloading configuration for the kubelet from the "kubelet-config-1.14" ConfigMap in the kube-system namespace
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Activating the kubelet service
[kubelet-start] Waiting for the kubelet to perform the TLS Bootstrap...

This node has joined the cluster:
* Certificate signing request was sent to apiserver and a response was received.
* The Kubelet was informed of the new secure connection details.

Run 'kubectl get nodes' on the control-plane to see this node join the cluster.
[k8smaster1@k8smaster1 ~]$ kubectl get nodes
NAME         STATUS     ROLES    AGE   VERSION
k8smaster1   Ready      master   42m   v1.14.3
k8snode1     Ready      <none>   12s   v1.14.3

I want to test if the node can be joined by remote.  so I use ddns service.
192.168.88.131:6443 = weinformatics.vicp.io:36791. It can not finished and stop at this status:
[root@k8snode1 ~]# kubeadm join weinformatics.vicp.io:36791 --token idiny6.30idongoafemg1bp \
> --discovery-token-ca-cert-hash sha256:a9ac241ea191d51a99c311ba57265eeff6733298530243c1a0b1e143062b9e08
[preflight] Running pre-flight checks

It seems that i got the same info by 192.168.88.131:6443 or weinformatics.vicp.io:36791
[k8snode1@k8snode1 ~]$ curl https://192.168.88.131:6443
curl: (60) Peer's Certificate issuer is not recognized.
More details here: http://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
[k8snode1@k8snode1 ~]$ 

[k8snode1@k8snode1 ~]$ curl https://weinformatics.vicp.io:36791
curl: (60) Peer's Certificate issuer is not recognized.
More details here: http://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
[k8snode1@k8snode1 ~]$

**Comment by neolit123** [comment]
is it stopping at?
> [preflight] Running pre-flight checks

try passing `--v=3` to `kubeadm join`

**Comment by cailiangliang765** [comment]
> is it stopping at?
> 
> > [preflight] Running pre-flight checks
> 
> try passing `--v=3` to `kubeadm join`

@neolit123 Many thanks for your reply ! It stop at [preflight] Running pre-flight checks for one hour so i have to terminate it. I add --v=3 to kubeadm join as you say.

[root@k8snode1 ~]# kubeadm join weinformatics.vicp.io:36791 --v=3 --token idiny6.30idongoafemg1bp \
> --discovery-token-ca-cert-hash sha256:a9ac241ea191d51a99c311ba57265eeff6733298530243c1a0b1e143062b9e08
I0613 23:09:05.554492   22036 join.go:367] [preflight] found NodeName empty; using OS hostname as NodeName
I0613 23:09:05.554571   22036 initconfiguration.go:105] detected and using CRI socket: /var/run/dockershim.sock
[preflight] Running pre-flight checks
I0613 23:09:05.554652   22036 preflight.go:90] [preflight] Running general checks
I0613 23:09:05.554701   22036 checks.go:254] validating the existence and emptiness of directory /etc/kubernetes/manifests
I0613 23:09:05.554759   22036 checks.go:292] validating the existence of file /etc/kubernetes/kubelet.conf
I0613 23:09:05.554768   22036 checks.go:292] validating the existence of file /etc/kubernetes/bootstrap-kubelet.conf
I0613 23:09:05.554776   22036 checks.go:105] validating the container runtime
I0613 23:09:05.686241   22036 checks.go:131] validating if the service is enabled and active
I0613 23:09:05.792610   22036 checks.go:341] validating the contents of file /proc/sys/net/bridge/bridge-nf-call-iptables
I0613 23:09:05.792681   22036 checks.go:341] validating the contents of file /proc/sys/net/ipv4/ip_forward
I0613 23:09:05.792714   22036 checks.go:653] validating whether swap is enabled or not
I0613 23:09:05.792795   22036 checks.go:382] validating the presence of executable ip
I0613 23:09:05.792817   22036 checks.go:382] validating the presence of executable iptables
I0613 23:09:05.792850   22036 checks.go:382] validating the presence of executable mount
I0613 23:09:05.792870   22036 checks.go:382] validating the presence of executable nsenter
I0613 23:09:05.792898   22036 checks.go:382] validating the presence of executable ebtables
I0613 23:09:05.794579   22036 checks.go:382] validating the presence of executable ethtool
I0613 23:09:05.794618   22036 checks.go:382] validating the presence of executable socat
I0613 23:09:05.794637   22036 checks.go:382] validating the presence of executable tc
I0613 23:09:05.794661   22036 checks.go:382] validating the presence of executable touch
I0613 23:09:05.794690   22036 checks.go:524] running all checks
I0613 23:09:05.828014   22036 checks.go:412] checking whether the given node name is reachable using net.LookupHost
I0613 23:09:05.828200   22036 checks.go:622] validating kubelet version
I0613 23:09:05.956281   22036 checks.go:131] validating if the service is enabled and active
I0613 23:09:05.965707   22036 checks.go:209] validating availability of port 10250
I0613 23:09:05.965916   22036 checks.go:292] validating the existence of file /etc/kubernetes/pki/ca.crt
I0613 23:09:05.965932   22036 checks.go:439] validating if the connectivity type is via proxy or direct
I0613 23:09:05.965967   22036 join.go:427] [preflight] Discovering cluster-info
I0613 23:09:05.966090   22036 token.go:200] [discovery] Trying to connect to API Server "weinformatics.vicp.io:36791"
I0613 23:09:05.967252   22036 token.go:75] [discovery] Created cluster-info discovery client, requesting info from "https://weinformatics.vicp.io:36791"
I0613 23:09:06.555046   22036 token.go:141] [discovery] Requesting info from "https://weinformatics.vicp.io:36791" again to validate TLS against the pinned public key
I0613 23:09:06.730189   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:11.949959   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:16.904783   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:21.921743   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:26.913917   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:31.928859   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:36.917578   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:41.923710   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:46.909015   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:51.909845   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:09:56.921216   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
I0613 23:10:01.916372   22036 token.go:147] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io]
^C
[root@k8snode1 ~]#

**Comment by neolit123** [comment]
thanks for the log.

a bug on the kubeadm is that we should timeout and not wait for an hour.
i will file a separate issue for that.

------

> [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io

the problem here is that you are not adding `weinformatics.vicp.io` to the known domains in your API server certificate.

see these docs on how to pass your domian under `apiServer/certSANS` in the kubeadm v1beta1 config:

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/
https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#ClusterConfiguration

**Comment by cailiangliang765** [comment]
> apiServer/certSANS

I will need some time to finish this. Thank you !

**Comment by cailiangliang765** [comment]
> thanks for the log.
> 
> a bug on the kubeadm is that we should timeout and not wait for an hour.
> i will file a separate issue for that.
> 
> > [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io
> 
> the problem here is that you are not adding `weinformatics.vicp.io` to the known domains in your API server certificate.
> 
> see these docs on how to pass your domian under `apiServer/certSANS` in the kubeadm v1beta1 config:
> 
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/
> https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#ClusterConfiguration

@neolit123 I will need some time to finish this. Thank you !

**Comment by cailiangliang765** [comment]
> thanks for the log.
> 
> a bug on the kubeadm is that we should timeout and not wait for an hour.
> i will file a separate issue for that.
> 
> > [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io
> 
> the problem here is that you are not adding `weinformatics.vicp.io` to the known domains in your API server certificate.
> 
> see these docs on how to pass your domian under `apiServer/certSANS` in the kubeadm v1beta1 config:
> 
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/
> https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#ClusterConfiguration

@neolit123 I met a new problem as follows:
[root@k8snode1 ~]# kubeadm join weinformatics.uicp.io:50437 --v=3 --token 5q0zal.g0vv13dwkovjconj \
> --discovery-token-ca-cert-hash sha256:d7406a6794d4844a21dfc746ca22c92520d70f829374542ea9bccc5cdb78eba6
I0615 10:50:47.267559   21771 join.go:367] [preflight] found NodeName empty; using OS hostname as NodeName
I0615 10:50:47.267636   21771 initconfiguration.go:105] detected and using CRI socket: /var/run/dockershim.sock
[preflight] Running pre-flight checks
I0615 10:50:47.267705   21771 preflight.go:90] [preflight] Running general checks
I0615 10:50:47.267742   21771 checks.go:254] validating the existence and emptiness of directory /etc/kubernetes/manifests
I0615 10:50:47.267787   21771 checks.go:292] validating the existence of file /etc/kubernetes/kubelet.conf
I0615 10:50:47.267795   21771 checks.go:292] validating the existence of file /etc/kubernetes/bootstrap-kubelet.conf
I0615 10:50:47.267802   21771 checks.go:105] validating the container runtime
I0615 10:50:47.359045   21771 checks.go:131] validating if the service is enabled and active
I0615 10:50:47.438156   21771 checks.go:341] validating the contents of file /proc/sys/net/bridge/bridge-nf-call-iptables
I0615 10:50:47.438221   21771 checks.go:341] validating the contents of file /proc/sys/net/ipv4/ip_forward
I0615 10:50:47.438252   21771 checks.go:653] validating whether swap is enabled or not
I0615 10:50:47.438300   21771 checks.go:382] validating the presence of executable ip
I0615 10:50:47.438321   21771 checks.go:382] validating the presence of executable iptables
I0615 10:50:47.438371   21771 checks.go:382] validating the presence of executable mount
I0615 10:50:47.438382   21771 checks.go:382] validating the presence of executable nsenter
I0615 10:50:47.438403   21771 checks.go:382] validating the presence of executable ebtables
I0615 10:50:47.438424   21771 checks.go:382] validating the presence of executable ethtool
I0615 10:50:47.438442   21771 checks.go:382] validating the presence of executable socat
I0615 10:50:47.438455   21771 checks.go:382] validating the presence of executable tc
I0615 10:50:47.438498   21771 checks.go:382] validating the presence of executable touch
I0615 10:50:47.438519   21771 checks.go:524] running all checks
I0615 10:50:47.468112   21771 checks.go:412] checking whether the given node name is reachable using net.LookupHost
I0615 10:50:47.468249   21771 checks.go:622] validating kubelet version
I0615 10:50:47.526606   21771 checks.go:131] validating if the service is enabled and active
I0615 10:50:47.533998   21771 checks.go:209] validating availability of port 10250
I0615 10:50:47.534646   21771 checks.go:292] validating the existence of file /etc/kubernetes/pki/ca.crt
I0615 10:50:47.534665   21771 checks.go:439] validating if the connectivity type is via proxy or direct
I0615 10:50:47.534698   21771 join.go:427] [preflight] Discovering cluster-info
I0615 10:50:47.535052   21771 token.go:200] [discovery] Trying to connect to API Server "weinformatics.uicp.io:50437"
I0615 10:50:47.535668   21771 token.go:75] [discovery] Created cluster-info discovery client, requesting info from "https://weinformatics.uicp.io:50437"
I0615 10:50:57.590818   21771 token.go:83] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.uicp.io:50437/api/v1/namespaces/kube-public/configmaps/cluster-info: net/http: TLS handshake timeout]
I0615 10:51:12.653251   21771 token.go:83] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.uicp.io:50437/api/v1/namespaces/kube-public/configmaps/cluster-info: net/http: TLS handshake timeout]
I0615 10:51:27.647134   21771 token.go:83] [discovery] Failed to request cluster info, will try again: [Get https://weinformatics.uicp.io:50437/api/v1/namespaces/kube-public/configmaps/cluster-info: net/http: TLS handshake timeout]

**Comment by cailiangliang765** [comment]
> thanks for the log.
> 
> a bug on the kubeadm is that we should timeout and not wait for an hour.
> i will file a separate issue for that.
> 
> > [Get https://weinformatics.vicp.io:36791/api/v1/namespaces/kube-public/configmaps/cluster-info: x509: certificate is valid for k8smaster1, kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, not weinformatics.vicp.io
> 
> the problem here is that you are not adding `weinformatics.vicp.io` to the known domains in your API server certificate.
> 
> see these docs on how to pass your domian under `apiServer/certSANS` in the kubeadm v1beta1 config:
> 
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/
> https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#ClusterConfiguration

@neolit123 Hi, I add the apiServer/certSANS and update the as your refer sites, it works for me.
apiVersion: kubeadm.k8s.io/v1beta1
kind: ClusterConfiguration
kubernetesVersion: v1.14.3
apiServer:
  certSANs:
  - "weinformatics.uicp.io"
networking:
  dnsDomain: cluster.local
  podSubnet: 10.244.0.0/16
  serviceSubnet: 10.1.0.0/16

but another problems occurs, I known the reason was i can not use internal IP in the remote work node, as in "error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get config map: Get https://192.168.220.132:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: dial tcp 192.168.220.132:6443: i/o timeout". but i do not which file i have to change on master node or work node. Here is the log info:
[root@k8snode1 k8snode1]# kubeadm join weinformatics.uicp.io:50437 --v=3 --token 5q0zal.g0vv13dwkovjconj \
> --discovery-token-ca-cert-hash sha256:d7406a6794d4844a21dfc746ca22c92520d70f829374542ea9bccc5cdb78eba6
I0615 11:17:50.958859   26944 join.go:367] [preflight] found NodeName empty; using OS hostname as NodeName
I0615 11:17:50.958926   26944 initconfiguration.go:105] detected and using CRI socket: /var/run/dockershim.sock
[preflight] Running pre-flight checks
I0615 11:17:50.958984   26944 preflight.go:90] [preflight] Running general checks
I0615 11:17:50.959018   26944 checks.go:254] validating the existence and emptiness of directory /etc/kubernetes/manifests
I0615 11:17:50.959063   26944 checks.go:292] validating the existence of file /etc/kubernetes/kubelet.conf
I0615 11:17:50.959070   26944 checks.go:292] validating the existence of file /etc/kubernetes/bootstrap-kubelet.conf
I0615 11:17:50.959078   26944 checks.go:105] validating the container runtime
I0615 11:17:51.040247   26944 checks.go:131] validating if the service is enabled and active
I0615 11:17:51.110154   26944 checks.go:341] validating the contents of file /proc/sys/net/bridge/bridge-nf-call-iptables
I0615 11:17:51.110197   26944 checks.go:341] validating the contents of file /proc/sys/net/ipv4/ip_forward
I0615 11:17:51.110226   26944 checks.go:653] validating whether swap is enabled or not
I0615 11:17:51.110263   26944 checks.go:382] validating the presence of executable ip
I0615 11:17:51.110284   26944 checks.go:382] validating the presence of executable iptables
I0615 11:17:51.110314   26944 checks.go:382] validating the presence of executable mount
I0615 11:17:51.110323   26944 checks.go:382] validating the presence of executable nsenter
I0615 11:17:51.110363   26944 checks.go:382] validating the presence of executable ebtables
I0615 11:17:51.110496   26944 checks.go:382] validating the presence of executable ethtool
I0615 11:17:51.110517   26944 checks.go:382] validating the presence of executable socat
I0615 11:17:51.110531   26944 checks.go:382] validating the presence of executable tc
I0615 11:17:51.110553   26944 checks.go:382] validating the presence of executable touch
I0615 11:17:51.110570   26944 checks.go:524] running all checks
I0615 11:17:51.137771   26944 checks.go:412] checking whether the given node name is reachable using net.LookupHost
I0615 11:17:51.137956   26944 checks.go:622] validating kubelet version
I0615 11:17:51.201721   26944 checks.go:131] validating if the service is enabled and active
I0615 11:17:51.209662   26944 checks.go:209] validating availability of port 10250
I0615 11:17:51.209797   26944 checks.go:292] validating the existence of file /etc/kubernetes/pki/ca.crt
I0615 11:17:51.209810   26944 checks.go:439] validating if the connectivity type is via proxy or direct
I0615 11:17:51.209837   26944 join.go:427] [preflight] Discovering cluster-info
I0615 11:17:51.209896   26944 token.go:200] [discovery] Trying to connect to API Server "weinformatics.uicp.io:50437"
I0615 11:17:51.210380   26944 token.go:75] [discovery] Created cluster-info discovery client, requesting info from "https://weinformatics.uicp.io:50437"
I0615 11:17:51.723567   26944 token.go:141] [discovery] Requesting info from "https://weinformatics.uicp.io:50437" again to validate TLS against the pinned public key
I0615 11:17:52.404026   26944 token.go:164] [discovery] Cluster info signature and contents are valid and TLS certificate validates against pinned roots, will use API Server "weinformatics.uicp.io:50437"
I0615 11:17:52.404061   26944 token.go:206] [discovery] Successfully established connection with API Server "weinformatics.uicp.io:50437"
I0615 11:17:52.404097   26944 join.go:441] [preflight] Fetching init configuration
I0615 11:17:52.404122   26944 join.go:474] [preflight] Retrieving KubeConfig objects
[preflight] Reading configuration from the cluster...
[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get config map: Get https://192.168.220.132:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: dial tcp 192.168.220.132:6443: i/o timeout
[root@k8snode1 k8snode1]#

**Comment by neolit123** [comment]
> error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get config map: Get https://192.168.220.132:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: dial tcp 192.168.220.132:6443: i/o timeout

- do you have connectivity from the worker node the control-plane node?
- are you joining immediately after init - e.g. way too fast?

**Comment by cailiangliang765** [comment]
> > error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get config map: Get https://192.168.220.132:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: dial tcp 192.168.220.132:6443: i/o timeout
> 
> * do you have connectivity from the worker node the control-plane node?
> * are you joining immediately after init - e.g. way too fast?

1. In my master node, the weinformatics.uicp.io:50437 = 192.168.220.132:6443 by ddns service. The log shows I0615 11:17:52.404061 26944 token.go:206] [discovery] Successfully established connection with API Server "weinformatics.uicp.io:50437".  I do not know how to test the connectivity from the worker node the control-plane node.
2. I have three nodes: master node, worker node A are in the same LAN. and the worker node B was remote. In this issue, it was worker node B. I test master node and worker node A succesfully by kubectl get nodes with a satus tag (READY)

**Comment by neolit123** [comment]
not sure what the issue is here.

> dial tcp 192.168.220.132:6443: i/o timeout

suggests connectivity problems from the joining worker.

> the worker node B was remote. In this issue, it was worker node B. 

if B is a remote worker, why does it try to connect to a local address?
for the worker node "weinformatics.uicp.io:50437" should resolve to a remote address.

**Comment by cailiangliang765** [comment]
> not sure what the issue is here.
> 
> > dial tcp 192.168.220.132:6443: i/o timeout
> 
> suggests connectivity problems from the joining worker.
> 
> > the worker node B was remote. In this issue, it was worker node B.
> 
> if B is a remote worker, why does it try to connect to a local address?
> for the worker node "weinformatics.uicp.io:50437" should resolve to a remote address.

master node (192.168.220.132), worker node A (192.168.220.140) and B (192.168.88.132) That was also my confuse. I do not know which parameter on master or worker B node I need to edit to let dial tcp 192.168.220.132:6443: i/o timeout change to dial tcp weinformatics.uicp.io:50437: i/o timeout.

**Comment by neolit123** [comment]
i still do not understand the exact details here, but i will try to answer as much as i can.

please see what documentation updates this PR made:
https://github.com/kubernetes/website/pull/15524
(but these changes will appear in the 1.16 docs)

using `--apiserver-advertise-address=192.168.88.131` writes a serving certificate for the API server that allows `192.168.88.131` if you try to use anything other than `192.168.88.131` it will fail so you need to add values for the api server:

the best way to setup the cluster is to use a FQDN for controlPlaneEndpoint field of the kubeadm config and then join nodes to this FQDN instead of the address

https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2
```
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
controlPlaneEndpoint: "weinformatics.uicp.io:50437"
```
`--control-plane-endpoint` is available as a flag for kubeadm init in 1.15, but not earlier version so you have to use the config.

you can then map `weinformatics.uicp.io` to any IP using your DNS server, so your workers can join the cluster using that instead of the api-server-advertise address.
